### PR TITLE
Cschanaj script to trivialize targets

### PR DESCRIPTION
**Pull request by Hainish**

### CODE REVIEW

File: utils/trivialize-rules/package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
     "graceful-fs": "^4.1.11",
     "highland": "^2.7.1",
     "regulex": "0.0.2",
+    "valid-url": "^1.0.9",
```
Comment (pipboy96) [inline utils/trivialize-rules/package.json]: No need for this import. Node already supports `URL` object. Use:

```js
const isValidUrl = url => {
  try {
    new URL(url)
    return true
  } catch (e) {
    return false
  }
}
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        if (err) {
+          reject(err);
+        }
+        resolve();
+      });
+    });
+  });
+
+  // use for-loop to await too many file opened error
+  for (let fp of filePromises) {
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
  for (const fp of filePromises) {
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+        [, indent] = matches;
+        content = content.replace(regex, '\n');
+      }
+    }
+
+    return new Promise((resolve, reject) => {
+      fs.writeFile(path.join(rulesDir, filename), content, 'utf8', err => {
+        if (err) {
+          reject(err);
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
          reject(err);
          return
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+      return;
+    }
+
+    // (3) We check to make sure all targets are covered by rewrites
+    //     (a) if true, continue to (4)
+    //     (b) if some targets are not covered by any rewrites, this
+    //         doesn't affect our works trivializing the targets, but
+    //         we should give a warning and then proceed to (4)
+    targets.forEach(target => {
+      let supportedExplodedDomains = targetToSupportedExplodedDomainsMap.get(
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
      const supportedExplodedDomains = targetToSupportedExplodedDomainsMap.get(
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        }
+      }
+      unsupportedExplodedDomains.add(domain);
+      return false;
+    }
+
+    // initially, assume each target doesn't support any exploded domain
+    targets.forEach(target =>
+      targetToSupportedExplodedDomainsMap.set(target, [])
+    );
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]: Maybe replace this with `for...of`?

File: utils/trivialize-rules/trivialize-targets.js [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        // do not map non-wildcard targets to exploded domains
+        // otherwise, it will introduce unnecessary rewrites
+        targetToSupportedExplodedDomainsMap.delete(domain);
+        return true;
+      }
+
+      // this part follows the implementation in rules.js
+      let segments = domain.split('.');
+      for (let i = 1; i <= segments.length - 2; ++i) {
+        let tmp = '*.' + segments.slice(i, segments.length).join('.');
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
        const tmp = '*.' + segments.slice(i, segments.length).join('.');
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+      // make sure each exploded URL is valid
+      if (!explodedUrls.every(url => validUrl.isUri(url))) {
+        return false;
+      }
+
+      let isSimpleToAllExplodedUrls = true;
+      let isSnappingToSomeExplodedUrls = false;
+
+      for (let url of explodedUrls) {
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
      for (const url of explodedUrls) {
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        let { protocol, hostname, pathname } = new URL(url);
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
        const { protocol, hostname, pathname } = new URL(url);
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+    let explodedDomains = new Set();
+    let unsupportedExplodedDomains = new Set();
+    let unusedTargets = new Set();
+
+    // (1) We check if all rules can be exploded to valid urls
+    //     (a) if true, continue to (2)
+    //     (b) if we cannot trivialize targets for this ruleset, skip
+    function isExplosiveRewrite(rule) {
+      let explodedUrls = new Array();
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
      const explodedUrls = [];
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    let ruleToIsSimpleMap = new Map();
+    let ruleToIsSnappingMap = new Map();
+
+    let targetToSupportedExplodedDomainsMap = new Map();
+
+    let explodedDomains = new Set();
+    let unsupportedExplodedDomains = new Set();
+    let unusedTargets = new Set();
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]: All these may be changed to `const` with no dire consequences.

File: utils/trivialize-rules/trivialize-targets.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+    let warn = createTag('WARN', chalk.yellow, console.warn);
+    let info = createTag('INFO', chalk.green, console.info);
+    let fail = createTag('FAIL', chalk.red, console.error);
+
+    let content = await readFile(path.join(rulesDir, filename), 'utf8');
+    let { ruleset } = await parseString(content);
+
+    let rules = ruleset.rule.map(rule => rule.$);
+    let targets = ruleset.target.map(target => target.$.host);
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]: Try to always prefer `const` if possible.

File: utils/trivialize-rules/trivialize-targets.js [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+let fs = require('graceful-fs');
+let readdir = util.promisify(fs.readdir);
+let readFile = util.promisify(fs.readFile);
+let parseString = util.promisify(xml2js.parseString);
+
+let chalk = require('chalk');
+let validUrl = require('valid-url');
+let escapeStringRegexp = require('escape-string-regexp');
+let { explodeRegExp, UnsupportedRegExp } = require('./explode-regexp');
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]: Try to always prefer `const` if possible.

File: utils/trivialize-rules/trivialize-targets.js [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+let path = require('path');
+let xml2js = require('xml2js');
+
+let fs = require('graceful-fs');
+let readdir = util.promisify(fs.readdir);
+let readFile = util.promisify(fs.readFile);
+let parseString = util.promisify(xml2js.parseString);
+
+let chalk = require('chalk');
+let validUrl = require('valid-url');
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
const isValidUrl = url => {
  try {
    new URL(url)
    return true
  } catch (e) {
    return false
  }
}
```

File: utils/trivialize-rules/trivialize-targets.js [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+        // we are getting into runtime error, log and exit
+        if (!(e instanceof UnsupportedRegExp)) {
+          fail(e.message);
+          process.exit(1);
+        }
+        return false;
+      }
+
+      // make sure each exploded URL is valid
+      if (!explodedUrls.every(url => validUrl.isUri(url))) {
```
Comment (pipboy96) [inline utils/trivialize-rules/trivialize-targets.js]:
```suggestion
      if (!explodedUrls.every(isValidUrl)) {
```

**Comment by Hainish** [comment]
Cherry-picked in https://github.com/EFForg/https-everywhere/commit/de1fa652762a4b8d2d7c0e23beef6e993e0f63de

**Pull request closed without merging**